internal/controllers/reconciliation/controller.go

```diff
@@ -229,40 +220,80 @@ func (c *Controller) reconcileResource(ctx context.Context, comp *apiv1.Composit
 		return true, nil
 	}
 
-	// Compute a merge patch
-	var updated *unstructured.Unstructured
-	var typed bool
-	var err error
-	if resource.Replace {
-		updated = resource.Unstructured()
-	} else {
-		updated, typed, err = resource.Merge(ctx, prev, current, c.discovery)
+	// Dry-run the update to see if it's needed
+	dryRun, err := c.update(ctx, res, current, true)
+	if err != nil {
+		return false, fmt.Errorf("dry-run applying update: %w", err)
+	}
+	if resource.Compare(dryRun, current) {
+		return false, nil // in sync
+	}
+
+	// When using server side apply, make sure we haven't lost any managedFields metadata.
+	// Eno should always remove fields that are no longer set by the synthesizer, even if another client messed with managedFields.
+	if current != nil && prev != nil && !c.disableSSA && !res.Replace {
+		dryRunPrev := prev.Unstructured()
+		err = c.upstreamClient.Patch(ctx, dryRunPrev, client.Apply, client.ForceOwnership, client.FieldOwner("eno"), client.DryRunAll)
 		if err != nil {
-			return false, fmt.Errorf("performing three-way merge: %w", err)
+			return false, fmt.Errorf("getting managed fields values for previous version: %w", err)
+		}
+		if !resource.CompareEnoManagedFields(dryRunPrev.GetManagedFields(), current.GetManagedFields()) && !resource.CompareEnoManagedFields(dryRun.GetManagedFields(), current.GetManagedFields()) {
+			all := resource.MergeEnoManagedFields(current.GetManagedFields(), dryRunPrev.GetManagedFields())
+			current.SetManagedFields(all)
+			err := c.upstreamClient.Update(ctx, current, client.FieldOwner("eno"))
+			if err != nil {
+				return false, fmt.Errorf("updating managed fields metadata: %w", err)
+			}
+			logger.V(0).Info("corrected drift in managed fields metadata")
+			return true, nil
 		}
 	}
-	if updated == nil {
-		logger.V(1).Info("skipping empty update")
+
+	// Do the actual non-dryrun update
+	reconciliationActions.WithLabelValues("apply").Inc()
```

Comment:
After non-error?

Reply:
This one was inconsistent before - all of the other operations increment before sending the request. I have no strong opinions about which way it should go, want me to move all of them after the response?

Reply:
Ah I didn't realize this was the only inconsistent one, I'm good with it then :)